README
# kerodon

Acceptance test framework for web applications

Based on a similar api to Capybara.

## Usage

FIXME: write

## TODO

### Tests
* Write a simple ring app to test against
* Can we borrow some of Capybara's tests?

### Usage
* Write this readme
* Write some doc-strings

## License

Copyright (C) 2012 Nelson Morris

Distributed under the Eclipse Public License, the same as Clojure.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.9.1]

### Changed

* Bump flatland dependency to 1.5.7 to work under JDK 11.

## [0.9.0] - 2017-09-25

### Changed

* Updated dependencies to latest stable versions
* Add type hints for kerodon.test/some-text?
* Bump Peridot version

## [0.8.0] - 2016-06-13

### Changed

* Bump to Clojure 1.8
* Use arbitrary selector in `fill-in`

### Fixed

* Fix build-url function


src/kerodon/test.clj
(ns kerodon.test
  (:require [clojure.string :as s]
            [net.cgrand.enlive-html :as enlive]
            [kerodon.impl :as impl]))

(defn has
  ([state form]
     (has state form nil))
  ([state form msg]
     (form state msg)))

(defmacro validate [comparator generator expected exp-msg]
  `(fn [state# msg#]
     (try (let [value# (~generator state#)]
            (clojure.test/do-report {:actual value#
                                     :type (if (~comparator value# ~expected)
                                             :pass
                                             :fail)
                                     :message msg#
                                     :expected (quote ~exp-msg)}))
          (catch java.lang.Throwable t#
            (clojure.test/do-report {:actual t#
                                     :type :error
                                     :message msg#
                                     :expected (quote ~exp-msg)})))
     state#))

(defn- collapse-extra-whitespace
  "Replace one or more consecutive whitespaces with a single space,
  i.e., unwraps all text onto a single line."
  [state]
  (-> (:enlive state)
      (enlive/texts)
      (s/join)
      (s/replace #"\s+" " ")))

(defmacro validate-text
  "Common macro for all text validation"
  [test comparator expected]
  `(validate ~comparator
             ~collapse-extra-whitespace
             ~expected
             (~test ~expected)))

(defn re-match? [s r]
  (not (nil? (re-matches (re-pattern r) s))))

(defn re-find? [s r]
  (not (nil? (re-find (re-pattern r) s))))

(defmacro regex? [expected]
  `(validate-text ~'regex? re-match? ~expected))

(defmacro some-regex? [expected]
  `(validate-text ~'some-regex? re-find? ~expected))

(defmacro text? [expected]
  `(validate-text ~'text? = ~expected))

(defmacro some-text? [expected]
  `(validate-text ~'some-text? #(.contains ^String %1 ^CharSequence %2) ~expected))

(defn submap?
  "Checks whether m contains all entries in sub."
  [^java.util.Map m ^java.util.Map sub]
  (.containsAll (.entrySet m) (.entrySet sub)))

(defmacro link? [text & [href]]
  (let [expected (if (nil? href)
                   (list 'link? text)
                   (list 'link? text href))]
    `(validate (fn [coll# search#] (some #(submap? %1 search#) coll#))
               #(map (fn [link#]
                       {:href (get-in link# [:attrs :href])
                        :text (apply str (enlive/texts (:content link#)))})
                     (enlive/select (:enlive %) [:a]))
               (cond
                (nil? ~href) {:text ~text}
                (= :href ~text) {:href ~href}
                :else {:text ~text :href ~href})
               ~expected)))

(defmacro heading? [expected]
  `(validate #(some (partial = %2) %1)
             #(map (fn [h#] (apply str (enlive/texts (:content h#))))
                   (concat
                     (enlive/select (:enlive %) [:h1])
                     (enlive/select (:enlive %) [:h2])
                     (enlive/select (:enlive %) [:h3])
                     (enlive/select (:enlive %) [:h4])
                     (enlive/select (:enlive %) [:h5])
                     (enlive/select (:enlive %) [:h6])))
             ~expected
             (~'heading? ~expected)))

(defmacro status? [expected]
  `(validate =
            (comp :status :response)
            ~expected
            (~'status? ~expected)))

(defmacro value? [selector expected]
  `(validate =
             #(impl/get-value % ~selector)
             ~expected
             (~'value? ~selector ~expected)))

(defmacro missing? [selector]
  `(validate =
             #(count (enlive/select (:enlive %) ~selector))
             0
             (~'missing? ~selector)))

(defmacro attr? [selector attr expected]
  `(validate =
             #(impl/get-attr % ~selector ~attr)
             ~expected
             (~'attr? ~selector ~attr ~expected)))


test/kerodon/test/core.clj
(ns kerodon.test.core
  (:use [kerodon.core]
        [clojure.test]
        [kerodon.test])
  (:require [net.cgrand.moustache :as moustache]
            [ring.util.response :as response]
            [ring.middleware.params :as params]
            [ring.middleware.session :as session]
            [ring.middleware.multipart-params :as multipart-params]
            [net.cgrand.enlive-html :as enlive]
            [hiccup.core :as hiccup]
            [clojure.java.io :as io]))

(def app
  (params/wrap-params
   (session/wrap-session
    (moustache/app
     [""]
     {:get (fn [{:keys [session]}]
             (if-let [user (:user session)]
               (response/response (str "hi " user))
               (response/response (hiccup/html [:a {:id "go-login" :href "/login"} "login"]))))}
     ["login"]
     {:get (constantly
            (response/response
             (hiccup/html [:form {:action "/login" :method "post"}
                           [:label {:for "user"} "User"]
                           [:input {:type "text" :name "user"}]
                           [:label {:for "password"} "Password"]
                           [:input {:type "password" :name "password"}]
                           [:input {:type "submit" :value "Login"}]])))
      :post (fn [{:keys [params]}]
              (if (and (= (params "user")
                          "someone")
                       (= (params "password")
                          "password"))
                (assoc (response/redirect "/") :session {:user "someone"})
                (response/response "Bad login")))}
     ["login-with-css"]
     {:get (constantly
            (response/response
             (hiccup/html [:form {:action "/login" :method "post"}
                           [:input {:id "user" :type "text" :name "user"}]
                           [:input {:class "password"
                                    :type "password"
                                    :name "password"}]
                           [:input {:id "some-button"
                                    :type "submit"
                                    :value "Login"}]])))}
     ["login-without-method"]
     {:get (constantly
            (response/response
             (hiccup/html [:form {:action "/login"}
                           [:label {:for "user"} "User"]
                           [:input {:type "text" :name "user"}]
                           [:label {:for "password"} "Password"]
                           [:input {:type "password" :name "password"}]
                           [:input {:type "submit" :value "Login"}]])))}
     ["login-without-action"]
     {:get (constantly
            (response/response
             (hiccup/html [:form {:method "post"}
                           [:label {:for "user"} "User"]
                           [:input {:type "text" :name "user"}]
                           [:label {:for "password"} "Password"]
                           [:input {:type "password" :name "password"}]
                           [:input {:type "submit" :value "Login"}]])))
      :post (fn [{:keys [params]}]
              (if (and (= (params "user")
                          "someone")
                       (= (params "password")
                          "password"))
                (assoc (response/redirect "/") :session {:user "someone"})
                (response/response "Bad login")))}
     ["login-with-text-area"]
     {:get (constantly
            (response/response
             (hiccup/html [:form {:action "/login" :method "post"}
                           [:label {:for "user"} "User"]
                           [:textarea {:name "user"}]
                           [:label {:for "password"} "Password"]
                           [:input {:type "password" :name "password"}]
                           [:input {:type "submit" :value "Login"}]])))}
     ["logout"]
     {:post (fn [{:keys [params]}]
              (assoc (response/redirect "/") :session nil))}
     ["login-x2"]
     {:get (constantly
            (response/response
             (hiccup/html [:div
                           [:form {:id "1"}
                            [:label {:for "user"} "User"]
                            [:input {:type "text" :name "user"}]
                            [:label {:for "password"} "Password"]
                            [:input {:type "password" :name "password"}]
                            [:input {:type "submit" :value "Login"}]]
                           [:form {:id "2"}
                            [:label {:for "user"} "User"]
                            [:input {:type "text" :name "user"}]
                            [:label {:for "password"} "Password"]
                            [:input {:type "password" :name "password"}]
                            [:input {:type "submit" :value "Login"}]]])))
      :post (constantly (response/response "hi"))}
     ["title"]
     {:get (constantly
            (response/response
             (hiccup/html [:html
                           [:head [:title "some title"]]
                           [:body]])))}))))

(def upload-app
  (multipart-params/wrap-multipart-params
   (params/wrap-params
    (moustache/app
     ["upload"]
     {:get (constantly
            (response/response
             (hiccup/html
              [:form {:enctype "multipart/form-data"}
               [:label {:for "file"} "File"]
               [:input {:id "file" :name "file" :type "file"}]
               [:input {:type "submit" :value "upload"}]])))
      :post (fn [{:keys [params body]}]
              (response/response
               (hiccup/html [:html
                             [:body
                              (when-let [file (:tempfile (params "file"))]
                                (slurp file))]])))}))))

(def upload-app-with-params
  (multipart-params/wrap-multipart-params
   (params/wrap-params
    (moustache/app
     ["upload"]
     {:get (constantly
            (response/response
             (hiccup/html
              [:form {:enctype "multipart/form-data"}
               [:label {:for "u"} "Name"]
               [:input {:type "text" :name "u"}]
               [:input {:id "file" :name "file" :type "file"}]
               [:input {:type "submit" :value "upload"}]])))
      :post (fn [{:keys [params body]}]
              (response/response
               (hiccup/html [:html
                             [:body
                              (str (params "u")
                                   " "
                                   (when-let [file (:tempfile (params "file"))]
                                     (slurp file)))]])))}))))

(deftest good-form
  (-> (session app)
      (visit "/")
      (follow "login")
      (fill-in "User" "someone")
      (fill-in "Password" "password")
      (press "Login")
      (follow-redirect)
      (has (text? "hi someone")
           "press sends form fields to action url")
      (has (missing? [:#go-login])
           "Signed in user should not see login link")))

(deftest selector-not-found
  (let [state (-> (session app)
                  (visit "/"))]
    (is (thrown-with-msg? Exception
          #"field could not be found with selector \"NonExistant\""
          (fill-in state "NonExistant" "")))
    (is (thrown-with-msg? Exception
          #"button could not be found with selector \"NonExistant\""
          (press state "NonExistant")))
    (is (thrown-with-msg? Exception
          #"link could not be found with selector \"NonExistant\""
          (follow state "NonExistant")))))

(deftest follow-by-css
  (-> (session app)
      (visit "/")
      (follow [:#go-login])
      (has (text? "UserPassword"))))

(deftest form-by-css
  (-> (session app)
      (visit "/login-with-css")
      (fill-in [:#user] "someone")
      (fill-in [:.password] "password")
      (press [:#some-button])
      (follow-redirect)
      (has (text? "hi someone")
           "fill-in and press can find by css")))

(deftest form-without-action
  (-> (session app)
      (visit "/login-without-action")
      (fill-in "User" "someone")
      (fill-in "Password" "password")
      (press "Login")
      (follow-redirect)
      (has (text? "hi someone")
           "press sends form fields to same url when action is blank")))

(deftest form-without-method
  (-> (session app)
      (visit "/login-without-method")
      (fill-in "User" "someone")
      (fill-in "Password" "password")
      (press "Login")
      (follow-redirect)
      (has (text? "hi someone")
           "press sends form fields to action url with post when method blank")))

(deftest text-area-can-be-filled
  (-> (session app)
      (visit "/login-with-text-area")
      (fill-in "User" "someone")
      (fill-in "Password" "password")
      (press "Login")
      (follow-redirect)
      (has (text? "hi someone")
           "fill-in and press work for textareas")))

(deftest fill-in-can-be-checked-with-value?
  (-> (session app)
      (visit "/login")
      (fill-in "User" "someone")
      (has (value? "User" "someone"))))

(deftest within-generic
  (-> (session app)
      (visit "/login-x2")
      (within [:#2]
              (fill-in "User" "someone"))
      (within [:#2]
              (has (value? "User" "someone")))
      (within [:#1]
              (has (value? "User" "")))
      (within [:#2]
              (press "Login"))
      (has (text? "hi"))))

(deftest status-generic
  (-> (session app)
      (visit "/login")
      (has (status? 200))
      (visit "/non-existant")
      (has (status? 404))))

(deftest title-generic
  (-> (session app)
      (visit "/title")
      (within [:title]
              (has (text? "some title")))))

(deftest test-attrs
  (-> (session app)
      (visit "/login-x2")
      (has (attr? [:form#1 :label] :for "user"))
      (within [:div]
              (has (attr? [:form] :id "1")))))

(deftest attach-file-alone
  (-> (session upload-app)
      (visit "/upload")
      (attach-file [:#file] (io/file (io/resource "file.txt")))
      (press "upload")
      (doto
          (#(is (re-find #"multipart/form-data;"
                         (:content-type (:request %))))))
      (has (text? "hi from uploaded file\n"))))

(deftest attach-file-by-label
  (-> (session upload-app)
      (visit "/upload")
      (attach-file "File" (io/file (io/resource "file.txt")))
      (press "upload")
      (doto
          (#(is (re-find #"multipart/form-data;"
                         (:content-type (:request %))))))
      (has (text? "hi from uploaded file\n"))))

(deftest attach-file-with-params
  (-> (session upload-app-with-params)
      (visit "/upload")
      (fill-in "Name" "name")
      (attach-file [:#file] (io/file (io/resource "file.txt")))
      (press "upload")
      (has (text? "name hi from uploaded file\n"))))

(deftest missing-selector
  (-> (session app)
      (visit "/login")
      (has (missing? [:#no-such-element]))))

(deftest missing-selector-within-scope
  (-> (session app)
      (visit "/login")
      (within [:body]
              (has (missing? [:title])))))


project.clj
(defproject kerodon "0.9.2-SNAPSHOT"
  :description "Acceptance test framework for web applications"
  :url "https://github.com/xeqi/kerodon"
  :min-lein-version "2.0.0"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [peridot "0.5.1"]
                 [enlive "1.1.6" :exclusions [org.clojure/clojure]]
                 [ring/ring-codec "1.0.1"]
                 [org.flatland/ordered "1.5.7"]]
  :release-tasks [["vcs" "assert-committed"]
                  ["change" "version" "leiningen.release/bump-version" "release"]
                  ["vcs" "commit"]
                  ["vcs" "tag" "v"]
                  ["deploy"]
                  ["change" "version" "leiningen.release/bump-version"]
                  ["vcs" "commit"]
                  ["vcs" "push"]]
  :deploy-repositories [["releases" :clojars]
                        ["snapshots" :clojars]]
  :profiles {:test {:dependencies [[net.cgrand/moustache "1.1.0"
                                    :exclusions
                                    [[org.clojure/clojure]
                                     [ring/ring-core]]]
                                   [ring/ring-core "1.5.0"]
                                   [javax.servlet/servlet-api "2.5"]
                                   [hiccup "1.0.5"]]
                    :resource-paths ["test-resources"]}
             :1.5  {:dependencies [^:replace [org.clojure/clojure "1.5.1"]]}
             :1.6  {:dependencies [^:replace [org.clojure/clojure "1.6.0"]]}
             :1.7  {:dependencies [^:replace [org.clojure/clojure "1.7.0"]]}
             :1.8  {:dependencies [^:replace [org.clojure/clojure "1.8.0"]]}}
  :aliases {"all" ["with-profile" "test,1.5:test,1.6:test,1.7:test,1.8"]})


src/kerodon/core.clj
(ns kerodon.core
  (:require [kerodon.impl :as impl]
            [peridot.core :as peridot]))

(def session peridot/session)

(defn- resolve-uri [state ^String uri]
  (if-let [request (:request state)]
    (str (.resolve (java.net.URI. (peridot.request/url request)) uri))
    uri))

(defn visit [state uri & rest]
  (impl/include-parse (apply peridot/request state (resolve-uri state uri) rest)))

(defn follow-redirect [state]
  (impl/include-parse (peridot/follow-redirect state)))

(defn follow [state selector]
  (visit state (impl/find-url state selector)))

(defn fill-in [state selector input]
  (impl/set-value state selector input))

(defn choose [state selector option]
  (impl/choose-value state selector option))

(defn check [state selector]
  (impl/check-box state selector))

(defn uncheck [state selector]
  (impl/uncheck-box state selector))

(defn attach-file [state selector file]
  (impl/set-value state selector file))

(defn press [state selector]
  (apply visit state (impl/build-request-details state selector)))

(defmacro within [state selector & fns]
  `(impl/using ~state ~selector #(-> % ~@fns)))


src/kerodon/impl/lookup.clj
(ns kerodon.impl.lookup
  (:require [net.cgrand.enlive-html :as enlive]
            [clojure.string :as string]))

(def fillable
  #{[:input
     (enlive/but
      (enlive/attr-has :type "submit"))]
    :textarea})

(defn css-or-content [selector]
  (if (string? selector)
    (enlive/pred #(= (:content %) [selector]))
    selector))

(defn find-form-with-submit [node selector]
  (enlive/select node
                 [[:form (enlive/has [[:input
                                       (enlive/attr= :type "submit")
                                       (if (string? selector)
                                         (enlive/attr= :value selector)
                                         selector)]])]]))

(defn find-submit [node text]
  (enlive/select node
                 [[:input
                   (enlive/attr= :type "submit")
                   (enlive/attr= :value text)]]))

(defn include-parse [state]
  (assoc state
    :enlive (when-let [body (:body (:response state))]
              (enlive/html-resource
               (java.io.StringReader. body)))))

;TODO: merge w/ peridot
(defn build-url
  [{:keys [scheme server-name port uri query-string]}]
  (str (name scheme)
       "://"
       server-name
       (when (and port
                  (not= port (scheme {:https 443 :http 80})))
         (str ":" port))
       uri
       query-string))

(defn find-form-element [node selector]
  (if-let [elem (first (enlive/select
                        node
                        [:form (css-or-content selector)]))]
    elem
    (throw (Exception.
            (str "field could not be found with selector \"" selector "\"")))))

(defn get-form-element-by-name [name]
  [:form (enlive/attr-has :name name)])

(defn name-from-element [elem]
  (if (= :input (:tag elem))
    (:name (:attrs elem))
    (:for (:attrs elem))))

(defn get-value [state selector]
  (-> (enlive/select (:enlive state)
                     (-> (find-form-element (:enlive state) selector)
                         name-from-element
                         get-form-element-by-name))
      first
      :attrs
      :value))

(defn set-value [state selector input]
  (update-in state [:enlive]
             (fn [node]
               (enlive/transform node
                                 (-> (find-form-element node selector)
                                     name-from-element
                                     get-form-element-by-name)
                                 (fn [snode]
                                   (assoc-in snode [:attrs :value] input))))))

(defn find-link [state selector]
  (if-let [link (first (enlive/select state [[:a (css-or-content selector)]]))]
    link
    (throw (Exception.
            (str "link could not be found with selector \""
                 selector "\"")))))

(defn find-url [state selector]
  (-> (:enlive state)
      (find-link selector)
      :attrs
      :href))

(defn using [state selector f]
  (let [new-state (f (update-in state [:enlive]
                                #(enlive/select
                                  %
                                  selector)))]
    (if (and (= (:request new-state)
                (:request state))
             (= (:response new-state)
                (:response state)))
      (update-in state
                 [:enlive]
                 #(enlive/transform %
                                    selector
                                    (constantly (:enlive new-state))))
      new-state)))

(defn build-request-details [state selector]
  (if-let [form (first (find-form-with-submit (:enlive state) selector))]
    (let [method (keyword (string/lower-case (:method (:attrs form) "post")))
          url (or (:action (:attrs form))
                  (build-url (:request state)))
          params (into {}
                       (map (comp (juxt (comp str :name)
                                        :value) :attrs)
                            (enlive/select form
                                           [fillable])))]
      [url :request-method method :params params])
    (throw (Exception.
            (str "button could not be found with selector \""
                 selector "\"")))))

src/kerodon/impl.clj
(ns kerodon.impl
  (:require [net.cgrand.enlive-html :as enlive]
            [clojure.string :as string]
            [clojure.java.io :as io]
            [flatland.ordered.map :as om])
  (:import java.io.StringReader))

(defn- assoc-conj
  "Associate a key with a value in a map. If the key already exists in the map,
  a vector of values is associated with the key.
  Stolen out of ring-codec."
  [map key val]
  (assoc map key
    (if-let [cur (get map key)]
      (if (vector? cur)
        (conj cur val)
        [cur val])
      val)))

;; selectors

(defn form-element-by
  "Turn a map of {:attr 'value'} into an elive selector for items
  in a form whose attributes all match"
  [attr-map]
  [:form (mapv (partial apply enlive/attr-has) attr-map)])

(defn css-or-content [selector]
  (if (string? selector)
    (enlive/pred #(let [fc (first (:content %))]
                    (and (string? fc)
                         (= (clojure.string/trim fc) selector))))
    selector))

(defn css-or-label [selector]
  (if (string? selector)
    (enlive/pred #(and (= (:tag %1) :label)
                       (= (-> (enlive/texts [%1]) first clojure.string/trim) selector)))
    selector))

(defn css-or-value [selector]
  (if (string? selector)
    (enlive/attr= :value selector)
    selector))

;; finders
(defn- not-found [type selector]
  (throw
    (IllegalArgumentException.
      (format "%s could not be found with selector \"%s\"" type selector))))

(defn form-and-button
  "Given a form, returns both the form and the specified submit button."
  [form selector]
  (let [query #{[:input
                  (enlive/attr= :type "submit")
                  (css-or-value selector)]
                 [:button
                  (enlive/attr= :type "submit")
                  (css-or-content selector)]}
        button (first (enlive/select form [query]))]
    [form button]))

(defn form-and-submit
  "Locates the form containing the specified submit button.
  Returns the form and the button itself."
  [node selector]
  (if-let [found
           (first (filter (fn [[form button]] (not (nil? button)))
                          (map (fn [form] (form-and-button form selector))
                               (enlive/select node [:form]))))]
    found
    (not-found "button" selector)))

(defn form-element [node selector]
  (if-let [elem (first (enlive/select
                        node
                        [:form (css-or-label selector)]))]
    elem
    (not-found "field" selector)))

(defn link [node selector]
  (if-let [link (first (enlive/select
                        node
                        [[:a (css-or-content selector)]]))]
    link
    (not-found "link" selector)))

(defn- field-to-selector [elem]
  (if-let [id (get-in elem [:attrs :id])]
    (form-element-by {:id id})
    (if (= "radio" (get-in elem [:attrs :type]))
      (form-element-by (select-keys (:attrs elem) [:name :value]))
      (form-element-by {:name (get-in elem [:attrs :name])}))))

(defn- label-to-selector [doc label]
  (if-let [id (get-in label [:attrs :for])]
    (let [selector (form-element-by {:id id})]
      (if (first (enlive/select doc selector))
        selector
        (not-found "field" (keyword (str "#" id)))))
    (if-let [field (first (enlive/select label [#{:input :select :textarea}]))]
      (field-to-selector field)
      (not-found "field inside label" (apply str (enlive/texts [label]))))))

(defn- form-element-selector [doc selector elem]
  (if (= :label (:tag elem))
    (label-to-selector doc elem)
    [selector]))

(defn form-element-query [node selector]
  (->> (form-element node selector)
       (form-element-selector node selector)))

(defn form-element-for [node selector]
  (enlive/select node (form-element-query node selector)))

;; state manipulation


(defprotocol ToHtmlResource
  (to-html-resource [obj]))

(extend-protocol ToHtmlResource
  clojure.lang.ISeq
  (to-html-resource [s]  (to-html-resource (apply str s)))
  String
  (to-html-resource [str] (enlive/html-resource (StringReader. str)))
  Object
  (to-html-resource [o] (enlive/html-resource o)))

(defn include-parse [state]
  (assoc state
    :enlive (when-let [body (:body (:response state))]
              (to-html-resource body))))

;TODO: merge w/ peridot
(defn build-url
  [{:keys [scheme server-name port uri query-string]}]
  (str (name scheme)
       "://"
       server-name
       (when (and port
                  (not= port (scheme {:https 443 :http 80})))
         (str ":" port))
       uri
       (when query-string
         (str "?" query-string))))

(defn get-value [state selector]
  (or (-> (:enlive state)
          (form-element-for selector)
          first
          :attrs
          :value)
      ""))

(defn get-attr [state selector attr]
  (-> (:enlive state)
      (enlive/select selector)
      first
      :attrs
      (get attr)))

(defn set-value [state selector input]
  (update-in state [:enlive]
             (fn [node]
               (enlive/transform node
                                 (form-element-query node selector)
                                 (fn [snode]
                                   (if (= (:tag snode) :textarea)
                                     (assoc-in snode [:content] [input])
                                     (assoc-in snode [:attrs :value] input)))))))

(defn- option=
  "Enlive selector for options with matching content or value"
  [option]
  [[:option #{(enlive/pred #(= [option] (:content %)))
              (enlive/attr= :value option)}]])

(defn- mark-option-selected [option]
  (fn [select-node]
    (when (empty? (enlive/select select-node (option= option)))
      (throw
        (IllegalArgumentException.
          (format "option could not be found with selector \"%s\"" option))))
    ; Transform requires a list of nodes, we only have one here
    (-> (list select-node)
        ; Deselect all options
        (enlive/transform [:option]
                          (enlive/remove-attr :selected))
        ; Select the chosen option
        (enlive/transform (option= option)
                          (enlive/set-attr :selected "selected")))))

(defn choose-value [state selector option]
  (let [enlive (:enlive state)]
    (assoc state :enlive
      (enlive/transform enlive
                        (form-element-query enlive selector)
                        (mark-option-selected option)))))

(defn check-box [state selector]
  (assoc state :enlive
    (enlive/transform (:enlive state)
                      (form-element-query (:enlive state) selector)
                      #(assoc-in % [:attrs :checked] "checked"))))

(defn uncheck-box [state selector]
  (assoc state :enlive
    (enlive/transform (:enlive state)
                      (form-element-query (:enlive state) selector)
                      #(update-in % [:attrs] dissoc :checked))))

(defn find-url [state selector]
  (-> (:enlive state)
      (link selector)
      :attrs
      :href))

(defn using [state selector f]
  (let [new-state (f (update-in state [:enlive]
                                #(enlive/select
                                  %
                                  selector)))]
    (if (and (= (:request new-state)
                (:request state))
             (= (:response new-state)
                (:response state)))
      (update-in state
                 [:enlive]
                 #(enlive/transform %
                                    selector
                                    (constantly (:enlive new-state))))
      new-state)))

;; Reading Form

(defmulti field->value
  "Get the value of the form field, or nil if it should be ignored"
  :tag)

(defmethod field->value :textarea
  ;; A textarea's child content is its value
  [field] (-> field :content first str))

(defn- selected-options
  "Get the <select> options that are considered selected"
  [options select-by-default]
  (or (seq (enlive/select options [(enlive/attr? :selected)]))
      (when select-by-default [(first options)])))

(defn- option-value
  "Get the real value of an <option> element"
  [option]
  (if-let [value (get-in option [:attrs :value])]
    value
    (first (:content option))))

(defmethod field->value :select
  ;; Single selects have the first element selected by default
  ;; Multiple selects do not have anything selected by default
  [field]
  (let [single (not (contains? (:attrs field) :multiple))]
    (-> (enlive/select field [:option])
        (selected-options single)
        (->> (map option-value))
        ((fn [values] (if single (first values) (seq values)))))))

(defn- is-submit-button?
  "Checks if the field is a form submit button."
  [field]
  (and (contains? #{:input :button} (:tag field))
       (= (get-in field [:attrs :type]) "submit")))

(defmethod field->value :button
  [field]
  (if (is-submit-button? field)
    (get-in field [:attrs :value])))

(declare input->value)
(defmethod field->value :input
  ;; Delegate input field handling to a second multimethod
  [field] (input->value field))

(defn- ns-keyword [name]
  (when name (keyword "kerodon.impl" name)))

(defmulti input->value
  "Get the value for each type of <input>, or nil if it should be ignored"
  #(-> % :attrs :type ns-keyword))

(derive ::checkbox ::box)
(derive ::radio ::box)

(defmethod input->value ::box
  ;;Radios and checkboxes only send their value when they are checked
  ;; The spec says no value is undefined, but de-facto standard is "on"
  [field]
  (when (contains? (:attrs field) :checked)
    (get-in field [:attrs :value] "on")))

(derive ::reset ::ignored)
(derive ::button ::ignored)

(defmethod input->value ::ignored
  [field]
  nil)

(defmethod input->value ::file
  ;; File inputs should not be coerced into strings
  [field]
  (get-in field [:attrs :value]))

(defmethod input->value :default
  ;; Any input not specified is treated as text input and stringified
  [field]
  (str (get-in field [:attrs :value])))

(defn field-name
  "Get the name attribute for a form field"
  [field] (get-in field [:attrs :name]))

(defn- other-submit-button?
  "is field a submit button other than the pressed one?"
  [field button]
  (and (not (nil? button))
       (is-submit-button? field)
       (not= button field)))

(defn all-form-params [form & [button]]
  (reduce
   (fn [params field]
     (if-not (other-submit-button? field button)
       (if-let [value (field->value field)]
         (assoc-conj params (field-name field) value)
         params)
       params))
   (om/ordered-map)
   (enlive/select form [[#{:input :textarea :select :button}
                         (enlive/but (enlive/attr? :disabled))
                         (enlive/attr? :name)]])))

(defn build-request-details [state selector]
  (let [[form button] (form-and-submit (:enlive state) selector)
        method (keyword (string/lower-case (:method (:attrs form) "get")))
        url (or (not-empty (:action (:attrs form)))
                (build-url (:request state)))
        params (all-form-params form button)]
    [url :request-method method :params params]))
